### PR TITLE
[Prompt registry] Phase 3 — Cutover: route default path through the registry

### LINKED ISSUE #3611: [Prompt registry] Phase 3 — Cutover: route default path through the registry
Part of the prompt-registry roadmap OpenHands/software-agent-sdk#3606 (proposal OpenHands/software-agent-sdk#2827).

**Goal:** make the registry the default production path, preserving every escape hatch and the cache contract.

**Depends on:** Phase 2 OpenHands/software-agent-sdk#3610.<br>**Blocks:** Phase 4 and Phase 5.

### PR
#3745 

### Tasks

- [x] Route `AgentBase.static_system_message` / `dynamic_context` (`agent/base.py:426,469`) through `PromptRegistry.build()` **for the default prompt only**.
- [x] Preserve every escape hatch: inline `system_prompt`, custom / absolute `system_prompt_filename`, `system_prompt_kwargs`, `security_policy_filename`. Custom-filename rendering still flows through `render_template`.
- [x] Leave `SystemPromptEvent` (`event/llm_convertible/system.py`) and `LLM._apply_prompt_caching` (`llm.py:2085`) untouched; `tests/sdk/llm/test_prompt_caching_cross_conversation.py` and the Phase 0 snapshot stay green.
- [x] Decision: optional transitional flag (e.g. `use_prompt_registry`) for canary rollout, or flip directly on the strength of the snapshot + caching tests.

### Acceptance criteria

- [x] Default agents render via the registry; all existing prompt / caching / agent-context tests green.
- [x] Inline-prompt and custom-template paths verified unchanged.

### PR DESCRIPTION
**Pull request by VascoSch92**
HUMAN:

Migration to new registry structure.

- [x] A human has tested these changes.

AGENT:

---

## Why

Part of the prompt-registry roadmap (#3606; proposal #2827), which replaces the
monolithic `system_prompt.j2` / `system_message_suffix.j2` + post-render `refine()`
with a typed section registry whose static/dynamic cache split is declared per-section
and unit-testable.

Phases 0–2 built and validated the registry **behind** the Phase 0 snapshot — nothing
was wired into the runtime. This is the **Phase 3 cutover** (#3611): it makes the
registry the default production path for assembling the system prompt, while preserving
every prompt-customization escape hatch and the cross-conversation caching contract.

## Summary

- Route the default prompt through the registry: `AgentBase.static_system_message` /
  `dynamic_context` (`agent/base.py`) and the production `Agent.get_dynamic_context(state)`
  (`agent/agent.py`) now assemble via `create_registry().build(ctx)` instead of
  `render_template` / `get_system_message_suffix`.
- Escape hatches preserved untouched: inline `system_prompt` returned verbatim; a
  custom/absolute `system_prompt_filename` still renders through Jinja (`render_template`,
  guarded by `!= "system_prompt.j2"`); `system_prompt_kwargs` + `security_policy_filename`
  flow into the registry context.
- **Direct flip, no transitional flag** (the decision in #3611) — backed by the Phase 2
  byte-for-byte equivalence tests plus the caching regression test.
- `SystemPromptEvent` and `LLM._apply_prompt_caching` left untouched; the
  cross-conversation caching test stays green unmodified (it pins constancy + the
  static/dynamic split, not exact bytes).
- Regenerated the Phase 0 snapshots for the one intentional effect: inter-section spacing
  collapses from the legacy template's 2–5 blank lines to the registry's single blank line
  — **whitespace only** (213 blank-line deletions, 0 content changes).
- Housekeeping: rename `default_registry.py` → `presets.py` and `build_default_registry()`
  → `create_registry()`; the module is now the home for named registry presets. A future
  `preset` flag on `create_registry()` will select among variants (interactive, planning,
  …) over the same engine; today it returns the default composition.

## Issue Number

Closes #3611 (roadmap #3606, proposal #2827). Depends on Phase 2 (#3610 / #3683).

## How to Test

Phase 3 wires the registry into the **runtime** path, so beyond the unit/equivalence
tests this is verified end-to-end: a real `Agent` runs `init_state()` and the emitted
`SystemPromptEvent`'s two content blocks are exactly what the registry produces.

```
uv run python - << 'EOF'
import uuid, tempfile
from openhands.sdk.agent import Agent
from openhands.sdk.context.agent_context import AgentContext
from openhands.sdk.context.prompts.presets import create_registry
from openhands.sdk.conversation import ConversationState
from openhands.sdk.event import SystemPromptEvent
from openhands.sdk.llm import LLM
from openhands.sdk.workspace import LocalWorkspace

tmp = tempfile.mkdtemp()
agent = Agent(llm=LLM(model="claude-sonnet-4-5", usage_id="e2e"), tools=[],
              agent_context=AgentContext(system_message_suffix="Follow repo conventions."))
state = ConversationState.create(id=uuid.uuid4(), workspace=LocalWorkspace(working_dir=tmp),
                                 persistence_dir=tmp + "/.state", agent=agent)
state.secret_registry.update_secrets({"GITHUB_TOKEN": "x"})
events = []
agent.init_state(state, on_event=events.append)
ev = next(e for e in events if isinstance(e, SystemPromptEvent))
ctx = agent._build_prompt_context(additional_secret_infos=state.secret_registry.get_secret_infos())
blocks = create_registry().build(ctx)
print("static  == registry.static :", ev.system_prompt.text == blocks.static)
print("dynamic == registry.dynamic:", ev.dynamic_context.text == blocks.dynamic)
EOF
# static  == registry.static : True
# dynamic == registry.dynamic: True
```

Equivalence + per-section unit tests (real `Agent` instances, not mocks — the registry
reproduces the legacy prompt; the only normalized difference is inter-section whitespace):

```
uv run pytest tests/sdk/context/prompts/ -q
# 146 passed
```

Caching contract + escape hatches + agent context (all unmodified, all green):

```
uv run pytest \
  tests/sdk/llm/test_prompt_caching_cross_conversation.py \
  tests/sdk/agent/test_system_prompt.py \
  tests/sdk/agent/test_build_prompt_context.py \
  tests/sdk/context/test_agent_context.py -q
# 73 passed
```

Snapshot regeneration is whitespace-only (no content drift):

```
git diff tests/sdk/context/prompts/snapshots/ | grep -E '^[-+]' \
  | grep -vE '^(\+\+\+|---)' | grep -cvE '^[-+]$'
# 0   (zero non-blank changed lines; 213 blank-line deletions across 28 snapshots)
```

Full lint/type gate:

```
uv run pre-commit run --files \
  openhands-sdk/openhands/sdk/context/prompts/presets.py \
  openhands-sdk/openhands/sdk/agent/base.py \
  openhands-sdk/openhands/sdk/agent/agent.py \
  tests/sdk/context/prompts/test_default_registry.py \
  tests/sdk/agent/test_build_prompt_context.py
# ruff format, ruff lint, pycodestyle, pyright, import rules, tool-registration — all pass
```

## Type

- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [ ] Breaking change
- [ ] Docs / chore

## Notes

- **Prompt output change in production is whitespace-only**: a single blank line between
  sections instead of the legacy template's 2–5. No semantic content change; the static
  block remains free of per-run/timestamp content (cache-safe).
- Cache contract is intentionally untouched: the static block is still constant across
  conversations and emitted as the cache-marked first content block; the dynamic block
  stays unmarked.
- The registry is rebuilt per `init_state` (twice per conversation, ~1.4 µs each) —
  measured ~2× faster than the Jinja path it replaces. Left un-memoized on purpose:
  `PromptRegistry` is mutable and a fresh instance is the documented override point.
- Unblocks Phase 4 (#3612) and Phase 5 (#3613).

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.13-nodejs22-slim` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.13-nodejs22-slim) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:0aa862a-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-0aa862a-python \
  ghcr.io/openhands/agent-server:0aa862a-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:0aa862a-golang-amd64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-golang-amd64
ghcr.io/openhands/agent-server:vasco-migration-prompt-golang-amd64
ghcr.io/openhands/agent-server:0aa862a-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:0aa862a-golang-arm64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-golang-arm64
ghcr.io/openhands/agent-server:vasco-migration-prompt-golang-arm64
ghcr.io/openhands/agent-server:0aa862a-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:0aa862a-java-amd64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-java-amd64
ghcr.io/openhands/agent-server:vasco-migration-prompt-java-amd64
ghcr.io/openhands/agent-server:0aa862a-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:0aa862a-java-arm64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-java-arm64
ghcr.io/openhands/agent-server:vasco-migration-prompt-java-arm64
ghcr.io/openhands/agent-server:0aa862a-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:0aa862a-python-amd64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-python-amd64
ghcr.io/openhands/agent-server:vasco-migration-prompt-python-amd64
ghcr.io/openhands/agent-server:0aa862a-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-amd64
ghcr.io/openhands/agent-server:0aa862a-python-arm64
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-python-arm64
ghcr.io/openhands/agent-server:vasco-migration-prompt-python-arm64
ghcr.io/openhands/agent-server:0aa862a-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-arm64
ghcr.io/openhands/agent-server:0aa862a-golang
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-golang
ghcr.io/openhands/agent-server:vasco-migration-prompt-golang
ghcr.io/openhands/agent-server:0aa862a-golang_tag_1.21-bookworm
ghcr.io/openhands/agent-server:0aa862a-java
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-java
ghcr.io/openhands/agent-server:vasco-migration-prompt-java
ghcr.io/openhands/agent-server:0aa862a-eclipse-temurin_tag_17-jdk
ghcr.io/openhands/agent-server:0aa862a-python
ghcr.io/openhands/agent-server:0aa862aa32d744eb69d86760c7b87df98f2c7544-python
ghcr.io/openhands/agent-server:vasco-migration-prompt-python
ghcr.io/openhands/agent-server:0aa862a-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `0aa862a-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `0aa862a-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by enyst** [comment]
@openhands tell me what this pr is doing and how. Post a comment.Then /codereview it, and post your feedback as review with gh api, you are free to approve it if it looks good. If you think it needs a small eval, you can say that.

**Comment by enyst** [comment]
@/tmp/pr3745_explanation_comment.md

**Review by enyst — COMMENTED** [review]
🟡 **Taste Rating**: Acceptable — the registry cutover is clean and well-tested, but one compatibility edge looks real.

**[CRITICAL ISSUES]**
- [openhands-sdk/openhands/sdk/agent/base.py, Line 479] **Backward Compatibility**: The default-name custom prompt-dir path appears to be dropped. `prompt_dir` is computed from the concrete agent class, and `system_prompt_filename` is documented as a relative filename loaded from the agent's prompts directory. Before this PR, a custom `Agent` subclass packaged with `prompts/system_prompt.j2` and leaving `system_prompt_filename` at the default would render that subclass template. Now the decision is based only on `system_prompt_filename != "system_prompt.j2"`, so that same subclass is forced through the OpenHands registry instead. I verified this with a temp subclass: `agent.prompt_dir` pointed at the subclass `prompts/` directory, but `static_system_message` returned the registry prompt. Please either preserve the Jinja path when `self.prompt_dir` is not the built-in OpenHands prompt dir, or explicitly document/remove this extension point with maintainer sign-off and a regression test for the supported escape hatches.

**[TESTING / VALIDATION]**
- Targeted local tests passed: `uv run pytest tests/sdk/context/prompts/test_default_registry.py tests/sdk/agent/test_system_prompt.py tests/sdk/agent/test_build_prompt_context.py -q` → 93 passed.
- This PR changes runtime prompt assembly / agent behavior, so per the repo-specific `/codereview` policy I am not approving without eval evidence. A small eval or human maintainer confirmation after lightweight benchmark validation would be prudent even if the compatibility issue above is addressed.

**[RISK ASSESSMENT]**
- [Overall PR] ⚠️ Risk Assessment: 🟡 MEDIUM

The core implementation is small and the tests are strong, but this is the production prompt path and therefore can affect agent behavior/evals. There is also a plausible public extension-point regression around custom agent prompt directories.

**VERDICT:**
❌ **Needs attention before approval**: fix or explicitly accept the custom `prompt_dir` behavior change, then get lightweight eval / maintainer confirmation for the prompt-behavior cutover.

**KEY INSIGHT:**
The registry cutover should key off “the built-in default prompt” rather than only the literal filename `system_prompt.j2`, because custom agents can have their own default-named template in their own prompt directory.

_This review was created by an AI agent (OpenHands) on behalf of the user._

---

> **Improve this review?** If any feedback above seems incorrect or irrelevant to this repository, you can teach the reviewer to do better:
>
> 1. Add a `.agents/skills/custom-codereview-guide.md` file to your branch (or edit it if one already exists) with the `/codereview` trigger and the context the reviewer is missing (e.g., "Security concerns about X do not apply here because Y"). See the [customization docs](https://docs.openhands.dev/openhands/usage/use-cases/code-review#customization) for the required frontmatter format.
> 2. Re-request a review - the reviewer reads guidelines from the PR branch, so your changes take effect immediately.
> 3. When your PR is merged, the guideline file goes through normal code review by repository maintainers.
>
> **Resolve with AI?** Install the [iterate skill](https://github.com/OpenHands/extensions/tree/main/skills/iterate) in your agent and run `/iterate` to automatically drive this PR through CI, review, and QA until it's merge-ready.
>
> Was this review helpful? React with 👍 or 👎 to give feedback.

File: openhands-sdk/openhands/sdk/agent/base.py [hunk 1]
```diff
@@ -456,20 +457,26 @@ def static_system_message(self) -> str:
...
-        )
+        # Custom / absolute filename: keep the Jinja render path (escape hatch).
+        if self.system_prompt_filename != "system_prompt.j2":
+            return render_template(
+                prompt_dir=self.prompt_dir,
+                template_name=self.system_prompt_filename,
+                **self._resolved_template_kwargs(),
+            )
+
+        return create_registry().build(self._build_prompt_context()).static
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/agent/base.py:494]: This condition preserves custom filenames, but not custom prompt directories that still use the default filename. `prompt_dir` is class-derived, so before this PR a subclass packaged with `prompts/system_prompt.j2` rendered that template through `render_template`; now it is routed to `create_registry()` purely because the filename is `system_prompt.j2`. Please gate the registry path on the built-in prompt directory too, or document that default-named subclass templates are no longer supported.

**Comment by VascoSch92** [comment]
@enyst the bot was correct :-) It is a corner-case but was missing. I fixed it.

**Review by juanmichelini — APPROVED** [review]
Smoke test with N = 5 worked fine. See https://openhands-eval-monitor.vercel.app/?run=swebench%2Flitellm_proxy-claude-sonnet-4-5-20250929%2F27649915647%2F&days=15

Sample test with N = 50 inferred fine, but failed to evaluate 
https://openhands-eval-monitor.vercel.app/?run=swebench%2Flitellm_proxy-claude-sonnet-4-5-20250929%2F27631198020%2F&days=15

it failed again on re evaluation 

https://openhands-eval-monitor.vercel.app/?run=swebench%2Flitellm_proxy-claude-sonnet-4-5-20250929%2F27638145378%2F&days=15

Approved, since it's probably noise, just mentioning it in case the error recurrs.
